File: InterpreterInGo/dart/lib/token.dart
abstract class Token {
  final String literal;
  Token(this.literal);
}

class Illegal extends Token {
  Illegal() : super("ILLEGAL");
}

class Plus extends Token {
  Plus() : super("+");
}

class Dash extends Token {
  Dash() : super("-");
}

class LeftParen extends Token {
  LeftParen() : super("(");
}

class RightParen extends Token {
  RightParen() : super(")");
}

class LeftBrace extends Token {
  LeftBrace() : super("{");
}

class RightBrace extends Token {
  RightBrace() : super("}");
}

class RightBracket extends Token {
  RightBracket() : super("]");
}

class LeftBracket extends Token {
  LeftBracket() : super("]");
}

class Comma extends Token {
  Comma() : super(",");
}

class SemiColon extends Token {
  SemiColon() : super(";");
}

class Colon extends Token {
  Colon() : super(":");
}

class Assign extends Token {
  Assign() : super("=");
}

class Number extends Token {
  Number(String literal) : super(literal);
}

class Identifier extends Token {
  Identifier(String literal) : super(literal);
}

class Func extends Token {
  Func(String literal) : super(literal);
}

class Let extends Token {
  Let(String literal) : super(literal);
}

class EndOfFile extends Token {
  EndOfFile() : super("EOF");
}

File: CraftingInterpreters/Kotlin/src/main/kotlin/lox/Token.kt
package lox

enum class TokenType{
    //Single char tokens
    LEFT_PAREN, RIGHT_PAREN, LEFT_BRACE, RIGHT_BRACE, COMMA,
    DOT, MINUS, PLUS, SEMICOLON, SLASH, STAR,

    //One or two character tokens
    BANG, BANG_EQUAL, EQUAL, EQUAL_EQUAL,
    GREATER, GREATER_EQUAL, LESS, LESS_EQUAL,

    //Literals
    IDENTIFIER, STRING, NUMBER,

    //Keywords
    AND, CLASS, ELSE, FALSE, FUN, FOR, IF, NIL, OR,
    PRINT, RETURN, SUPER, THIS, TRUE, VAR, WHILE,

    EOF
}
class Token(private val type: TokenType, val lexeme: String, val literal: Any?, val line: Int) {

    override fun toString(): String {
        return "$type $lexeme $literal"
    }
}